typing/typecore.ml

```diff
@@ -3919,38 +3919,51 @@ and type_expect_
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
   | Pexp_while(scond, sbody) ->
-      let cond =
-        type_expect (Env.add_region_lock env) (mode_var ()) scond
+      let cond_env,wh_cond_region =
+        if is_local_returning_expr scond then env, false
```

Comment:
The use of `is_local_returning_expr` means that the following is also a loop without a region:
```ocaml
while cond do
  print_endline "hello";
  incr counter;
  local_ ()
done
```
Is this intentional? It's a bit weird, but consistency with the rule for functions is good. (On the other hand, I'm not a huge fan of the deep inspection in the current rule for functions!)

Reply:
Yeah, this is weird, but after discussion with Leo I think the conclusion is that consistency mandates it.  In the future it would nice to have different keywords for some of the different uses for the current `local_` keyword.